filesystem/vdi.ksy

```diff
@@ -0,0 +1,228 @@
+meta:
+  id: vdi
+  endian: le
+  encoding: utf-8
+  application:
+    - VirtualBox
+    - QEMU
+    - VMWare Workstation
+  license: GPL3.0+
+doc-ref: https://github.com/qemu/qemu/blob/master/block/vdi.c
+#  - https://github.com/qemu/qemu/blob/master/block/vdi.c
+#  - https://www.virtualbox.org/browser/vbox/trunk/src/VBox/Storage/VDI.cpp
+#  - https://forums.virtualbox.org/viewtopic.php?t=8046
+seq:
+  - id: header
+    type: header
+
+instances:
+  block_discarded:
+    value: "0xfffffffe"
+  block_unallocated:
+    value: "0xffffffff"
+  blocks_map:
+    -orig-id: bmap
+    pos: header.bmap_offset
+    size: header.bmap_size
+    type: blocks_map
+    doc: >
+      block_index = offset_in_virtual_disk / block_size
+      actual_data_offset = blocks_map[block_index]*block_size+metadata_size+offset_in_block
+      
+      The bmap will take up blocks_in_image_max * sizeof(uint32_t) bytes;
+      since the bmap is read and written in a single operation, its size needs to be limited to INT_MAX; furthermore, when opening an image, the bmap size is rounded up to be aligned on BDRV_SECTOR_SIZE.
+      Therefore this should satisfy the following: blocks_in_image_max * sizeof(uint32_t) + BDRV_SECTOR_SIZE == INT_MAX + 1 (INT_MAX + 1 is the first value not representable as an int)
+      This guarantees that any value below or equal to the constant will, when multiplied by sizeof(uint32_t) and rounded up to a BDRV_SECTOR_SIZE boundary, still be below or equal to INT_MAX.
+  disk:
+    pos: header.blocks_offset
+    #size: header.header_main.disk_size
+    type: disk
+types:
+  header:
+    seq:
+      - id: text
+        type: str
+        size: 0x40
+
+      - id: signature
+        contents: [0x7F, 0x10, 0xDA, 0xBE]
+
+      - id: version
+        type: version
+
+      - id: header_size_optional
+        type: u4
+        if: subheader_size_is_dynamic
+        
+      - id: header_main
+        size: header_size
+        type: header_main
+    instances:
+      subheader_size_is_dynamic:
+        value: version.major>=1
+      header_size:
+        value: (subheader_size_is_dynamic?header_size_optional:336)
+      
+      bmap_size:
+        value: ((header_main.blocks_in_image * 4+header_main.geometry.sector_size-1) / header_main.geometry.sector_size)*header_main.geometry.sector_size
+      bmap_offset:
+        value: header_main.offset_bmap
+
+      blocks_offset:
+        -orig-id: data_offset
+        value: header_main.offset_data
```

Comment:
Again, it's really harsh. If it's "blocks", then it should be "blocks" everywhere.